qiskit/quantum_info/analysis/z2_symmetries.py

```diff
@@ -0,0 +1,471 @@
+        tol (float): The tolerance threshold for ignoring real and complex parts of a coefficient.
+
+    References:
+        [1]: Bravyi, S., et al, "Tapering off qubits to simulate fermionic Hamiltonians"
+            `arXiv:1701.08213 <https://arxiv.org/abs/1701.08213>`__
+    """
+
+    def __init__(
+        self,
+        symmetries: list[Pauli],
+        sq_paulis: list[Pauli],
+        sq_list: list[int],
+        tapering_values: Optional[list[int]] = None,
+        *,
+        tol: float = 1e-14,
+    ):
+        r"""
+        Args:
+            symmetries: The list of Pauli objects representing the $Z_2$ symmetries. These correspond to
+                the generators of the symmetry group $\langle \tau_1, \dots, \tau_k \rangle>$.
+            sq_paulis: The list of single - qubit Pauli $\sigma^x_{q(i)}$ anti-commuting with the
+                symmetry $\tau_i$ and commuting with all the other symmetries $\tau_{j\neq i}$.
+                These operators are used to construct the unitary Clifford operators.
+            sq_list: The list of indices $q(i)$ of the single-qubit Pauli operators used to build the
+                Clifford operators.
+            tapering_values: List of eigenvalues determining the symmetry sector for each symmetry.
+            tol: Tolerance threshold for ignoring real and complex parts of a coefficient.
+
+        Raises:
+            QiskitError: Invalid paulis. The lists of symmetries, single - qubit paulis support paulis
+            and tapering values must be of equal length. This length is the number of applied symmetries
+            and translates directly to the number of eliminated qubits.
+        """
+        if len(symmetries) != len(sq_paulis):
+            raise QiskitError(
+                f"The number of Z2 symmetries, {len(symmetries)}, has to match the number \
+                of single-qubit pauli operators, {len(sq_paulis)}."
+            )
+
+        if len(sq_paulis) != len(sq_list):
+            raise QiskitError(
+                f"The number of single-qubit pauli operators, {len(sq_paulis)}, has to match the length \
+                of the of single-qubit list, {len(sq_list)}."
+            )
+
+        if tapering_values is not None:
+            if len(sq_list) != len(tapering_values):
+                raise QiskitError(
+                    f"The length of the single-qubit list, {len(sq_list)}, must match the length of the \
+                    tapering values, {len(tapering_values)} ."
+                )
+
+        self._symmetries = symmetries
+        self._sq_paulis = sq_paulis
+        self._sq_list = sq_list
+        self.tapering_values = tapering_values
+        self.tol = tol
+
+    @property
+    def symmetries(self):
+        """Return symmetries."""
+        return self._symmetries
+
+    @property
+    def sq_paulis(self):
+        """Return sq paulis."""
+        return self._sq_paulis
+
+    @property
+    def cliffords(self) -> list[SparsePauliOp]:
```

Comment:
It might be nice to use [`quantum_info.operators.Clirfford`](https://github.com/Qiskit/qiskit-terra/blob/cace20b513fe7739a91e01547023a9855f780dbd/qiskit/quantum_info/operators/symplectic/clifford.py). It is easy to understand and may improve the performance of Clifford evolution.

Reply:
This is a good point but I wonder how to work aroundd that because SparsePauliOp do not have an `evolve()` method available. Would this mean working only with Paulis and PauliLists ?

Reply:
I see. The workaround is `SparsePauliOp.paulis.evolve`. Here, if the phase in `PauliList` changes, it must be multiplied to coefficients. The internal `PauliList` in `SparsePauliOp` assume that the phase is zero.

Can I implement the method to `SparsePauliOp`?

Reply:
I'm currently thinking of using the Clifford class but I am facing the following issue:

- The stabiliser theory tells us that there exist a Clifford operator U such that the symmetries satisfy $\tau_i = U \sigma_i^x U^+$ (dagger).

- It can be verified that the operator $U = \prod_i U_i$ with $U_i = 1/\sqrt{2} (\sigma_{q(i)} + \tau_i)$ satisfies this condition. We thus have an expression of this operator as a `SparsePauliOp` (as an operator)

- To build the Clifford operator, one needs some sort of `QuantumCircuit` (or equivalent), which we don't have access to.

So a conversion appears necessary at some point but I'm not sure how

Reply:
I feel that this can be kept for a later PR. What do you think?